engine/gamesys/src/gamesys/components/comp_factory.cpp

```diff
@@ -270,27 +304,88 @@ namespace dmGameSystem
             dmLogError("Trying to unload factory prototype resource while loading.");
             return false;
         }
-        if(component->m_Resource->m_Prototype)
+        if(resource->m_Prototype)
         {
-            dmResource::Release(dmGameObject::GetFactory(collection), component->m_Resource->m_Prototype);
-            component->m_Resource->m_Prototype = 0;
+            dmResource::Release(dmGameObject::GetFactory(collection), resource->m_Prototype);
+            resource->m_Prototype = 0;
         }
         return true;
     }
 
+    // scripting
+    dmResource::HFactory CompFactoryGetResourceFactory(FactoryWorld* world)
+    {
+        return world->m_Factory;
+    }
+
+    dmGameObject::HPrototype CompFactoryGetPrototype(dmGameObject::HCollection collection, FactoryComponent* component)
+    {
+        return GetPrototype(dmGameObject::GetFactory(collection), component);
+    }
+
+    const char* CompFactoryGetPrototypePath(FactoryComponent* component)
+    {
+        return CompFactoryGetResource(component)->m_PrototypePath;
+    }
+
     CompFactoryStatus CompFactoryGetStatus(FactoryComponent* component)
     {
         if(component->m_Loading)
         {
             return COMP_FACTORY_STATUS_LOADING;
         }
-        if(component->m_Resource->m_Prototype == 0x0)
+        if(CompFactoryGetResource(component)->m_Prototype == 0x0)
```

Comment:
While we cannot set the new prototype while it is loading, I still feel there are perhaps edge cases that are missed.

It is important to remember that the default resource isn't automatically unloaded when we set a new prototype.

Since the user can handle this themselves (get state, load, unload and set prototype), I feel the mechanics should work though.